node/network/approval-distribution/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+		let hashes: HashSet<Hash> = metas.iter().map(|m| m.hash.clone()).collect();
+		for meta in metas.iter() {
+			match self.blocks.entry(meta.hash.clone()) {
+				hash_map::Entry::Vacant(entry) => {
+					// TODO: can we include parent_hash in `BlockApprovalMeta`?
+					let parent_hash = match request_parent_hash(ctx, meta.hash.clone()).await {
+						Some(parent_hash) => parent_hash,
+						None => continue,
+					};
+
+					entry.insert(BlockEntry {
+						known_by: HashMap::new(),
+						number: meta.number,
+						parent_hash,
+						knowledge: Knowledge::default(),
+						candidates: HashMap::new(),
+					});
+				}
+				_ => continue,
+			}
+			// TODO: how do we make sure there are no duplicates?
+			self.blocks_by_number.entry(meta.number).or_default().push(meta.hash.clone());
+		}
+		for (peer_id, view) in self.peer_views.iter() {
+			let view_set = view.heads.iter().cloned().collect::<HashSet<_>>();
+			let intersection = view_set.intersection(&hashes);
+			let view_intersection = View {
+				heads: intersection.cloned().collect(),
+				finalized_number: view.finalized_number,
+			};
+			Self::unify_with_peer(
+				&mut self.blocks,
+				ctx,
+				metrics,
+				peer_id.clone(),
+				view_intersection,
+			).await;
+		}
+	}
+
+	async fn process_incoming_peer_message(
+		&mut self,
+		ctx: &mut impl SubsystemContext<Message = ApprovalDistributionMessage>,
+		metrics: &Metrics,
+		peer_id: PeerId,
+		msg: protocol_v1::ApprovalDistributionMessage,
+	) {
+		match msg {
+			protocol_v1::ApprovalDistributionMessage::Assignments(assignments) => {
+				tracing::trace!(
+					target: LOG_TARGET,
+					peer_id = %peer_id,
+					num = assignments.len(),
+					"Processing assignments from a peer",
+				);
+				// TODO: can we batch the circulation part?
+				for (assignment, claimed_index) in assignments.into_iter() {
+					self.import_and_circulate_assignment(
+						ctx,
+						metrics,
+						MessageSource::Peer(peer_id.clone()),
+						assignment,
+						claimed_index,
+					).await;
+				}
+			}
+			protocol_v1::ApprovalDistributionMessage::Approvals(approvals) => {
+				tracing::trace!(
+					target: LOG_TARGET,
+					peer_id = %peer_id,
+					num = approvals.len(),
+					"Processing approvals from a peer",
+				);
+				for approval_vote in approvals.into_iter() {
+					self.import_and_circulate_approval(
+						ctx,
+						metrics,
+						MessageSource::Peer(peer_id.clone()),
+						approval_vote,
+					).await;
+				}
+			}
+		}
+	}
+
+	async fn handle_peer_view_change(
+		&mut self,
+		ctx: &mut impl SubsystemContext<Message = ApprovalDistributionMessage>,
+		metrics: &Metrics,
+		peer_id: PeerId,
+		view: View,
+	) {
+		Self::unify_with_peer(&mut self.blocks, ctx, metrics, peer_id.clone(), view.clone()).await;
+		let finalized_number = view.finalized_number;
+		self.peer_views.insert(peer_id.clone(), view);
+
+		// cleanup
+		let blocks = &mut self.blocks;
+		self.blocks_by_number
+			.range(0..=finalized_number)
```

Comment:
From the guide: "We have to be on guard for peers doing crazy stuff like jumping their `finalized_number` forward 10 trillion blocks to try and get us stuck in a loop for ages".

If `BTreeMap::range` is resilient against that, then a comment here to that effect would be appreciated. This seems like one of the things to test.

Reply:
Also `last_finalized_number..=finalized_number`, maybe? Where that is gathered from the `.insert(..., view)` line above

Reply:
Yeah, we should add that check.